src/dataqueue/queue.cc

```diff
@@ -0,0 +1,1092 @@
+  // such as streams, may not be capable of preserving idempotency
+  // and therefore must not claim to be. If an entry claims to be
+  // idempotent and cannot preserve that quality, subsequent reads
+  // must fail with an error when a variance is detected.
+  bool is_idempotent() const override { return data_queue_->is_idempotent(); }
+
+  void MemoryInfo(node::MemoryTracker* tracker) const override {
+    tracker->TrackField("data_queue", data_queue_);
+  }
+
+  DataQueue& getDataQueue() { return *data_queue_; }
+
+  SET_MEMORY_INFO_NAME(DataQueueEntry);
+  SET_SELF_SIZE(DataQueueEntry);
+
+ private:
+  std::shared_ptr<DataQueue> data_queue_;
+
+  class ReaderImpl : public DataQueue::Reader,
+                     public std::enable_shared_from_this<ReaderImpl> {
+   public:
+    explicit ReaderImpl(std::shared_ptr<DataQueue::Reader> inner)
+        : inner_(std::move(inner)) {}
+
+    int Pull(DataQueue::Reader::Next next,
+             int options,
+             DataQueue::Vec* data,
+             size_t count,
+             size_t max_count_hint) override {
+      auto self = shared_from_this();
+      return inner_->Pull(
+          std::move(next), options, data, count, max_count_hint);
+    }
+
+    SET_NO_MEMORY_INFO()
+    SET_MEMORY_INFO_NAME(ReaderImpl)
+    SET_SELF_SIZE(ReaderImpl)
+
+   private:
+    std::shared_ptr<DataQueue::Reader> inner_;
+  };
+};
+
+// ============================================================================
+
+// An FdEntry reads from a file descriptor. A check is made before each read
+// to determine if the fd has changed on disc. This is a best-effort check
+// that only looks at file size, creation, and modification times. The stat
+// check is also async, so there's a natural race condition there where the
+// file could be modified between the stat and actual read calls. That's
+// a tolerable risk here. While FdEntry is considered idempotent, this race
+// means that it is indeed possible for multiple reads to return different
+// results if the file just happens to get modified.
+class FdEntry final : public EntryImpl {
+  // TODO(@jasnell, @flakey5):
+  // * This should only allow reading from regular files. No directories, no
+  // pipes, etc.
+  // * The reader should support accepting the buffer(s) from the pull, if any.
+  // It should
+  //   only allocate a managed buffer if the pull doesn't provide any.
+  // * We might want to consider making the stat on each read sync to eliminate
+  // the race
+  //   condition described in the comment above.
+ public:
+  static std::unique_ptr<FdEntry> Create(Environment* env, Local<Value> path) {
+    // We're only going to create the FdEntry if the file exists.
+    uv_fs_t req;
+    auto cleanup = OnScopeLeave([&] { uv_fs_req_cleanup(&req); });
+
+    auto buf = std::make_shared<BufferValue>(env->isolate(), path);
+    if (uv_fs_stat(nullptr, &req, buf->out(), nullptr) < 0) return nullptr;
```

Comment:
Yes, I know. Unfortunately using async i/o for this leads to a race condition that I'd rather avoid or at least minimize the likelihood of all together. I'll leave a TODO comment here to see if we can revisit this but for now I think I'll leave these as they are.

Reply:
I understand that, but still, `fs.openAsBlob()` cannot be synchronous if it is intended to be implemented properly at some point. That’s an issue with the API that _should_ be fixed even if the PR is currently intended to be experimental.

Reply:
I don't think it's feasible for this call to be asynchronous. The whole point is to verify if the file has changed before accessing it. Making it async would create a possible race condition from within Node.js itself. The only way to avoid this is to make the call sync.

This is not that bad overall because on modern HW statSync is extremely fast anyway. On real system it even take less time than the asynchronous version.

It seem a good tradeoff for the use cases that this API enables.

Reply:
> The whole point is to verify if the file has changed before accessing it.

Doing this is inherently subject to race conditions, regardless of how the stat call happens.

> Making it async would create a possible race condition from within Node.js itself.

That’s true regardless of whether this call is async.

>  The only way to avoid this is to make the call sync.

No, it doesn’t avoid this problem. You can only avoid race conditions if other modifying fs calls are also synchronized with this check, which they are not by default within a single Node.js process and are also not across multiple processes.

> This is not that bad overall because on modern HW statSync is extremely fast anyway.

It’s a good thing that Node.js by default doesn’t make many assumptions about hardware.

> On real system it even take less time than the asynchronous version.

That’s true for all asynchronous calls if you measure them in isolation.

> It seem a good tradeoff for the use cases that this API enables.

I’m a definite -1 on adding an API that locks the implementation into being synchronous if there’s no strong technical reason for making it synchronous. (In fact, this would be fairly trivial if this feature were implemented in JS instead of of C++, which considering the size of the added C++ code and the number of available C++ coders and reviewers in this project might have been a good idea anyway.)

Reply:
Very well. For now I'm going to leave the sync stat calls as is but I will have the `openAsBlob()` return a `Promise<Blob>` to give flexibility for future modifications.